zcash_client_backend/src/encoding.rs

```diff
@@ -26,15 +26,55 @@ where
     bech32::encode(hrp, data.to_base32(), Variant::Bech32).expect("hrp is invalid")
 }
 
-fn bech32_decode<T, F>(hrp: &str, s: &str, read: F) -> Result<Option<T>, Error>
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum KeyError {
```

Comment:
This should probably actually be `Bech32DecodeError`, as it is not specific to keys (it can be returned from parsing Sapling payment addresses). Similarly below the messages shouldn't refer to "key", but either "data" or "value".